modules/core/src/test/java/org/apache/ignite/plugin/PluginConfigurationTest.java

```diff
@@ -0,0 +1,196 @@
+
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.cluster.ClusterNode;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.internal.IgniteEx;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.jetbrains.annotations.Nullable;
+import org.junit.Test;
+
+/**
+ * Tests for Ignite plugin configuration.
+ */
+public class PluginConfigurationTest extends GridCommonAbstractTest {
+    /** Test plugin name. */
+    private static final String TEST_PLUGIN_NAME = "test_plugin";
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /**
+     * Tests that {@link ServiceLoader#load(Class)} result will be used if plugin providers is not configured
+     * explicitly.
+     */
+    @Test
+    public void testNullPluginProviders() throws Exception {
+        doTest(null, U.allPluginProviders());
+    }
+
+    /**
+     * Tests that {@link ServiceLoader#load(Class)} result will be used if plugin providers is empty for
+     * {@link IgniteConfiguration}.
+     */
+    @Test
+    @SuppressWarnings("ZeroLengthArrayAllocation")
+    public void testEmptyPluginProviders() throws Exception {
+        doTest(new PluginProvider[]{}, U.allPluginProviders());
+    }
+
+    /**
+     * Tests that explicitly configured plugin providers will be used.
+     */
+    @Test
+    public void testNotEmptyPluginProviders() throws Exception {
+        TestPluginProvider testPluginProvider = new TestPluginProvider();
+
+        doTest(new PluginProvider[]{testPluginProvider}, Collections.singletonList(testPluginProvider));
+    }
+
+    /**
+     * Asserts expectations.
+     *
+     * @param cfgProviders Config providers.
+     * @param expProviders Expected providers.
+     */
+    @SuppressWarnings("rawtypes")
+    private void doTest(PluginProvider<?>[] cfgProviders, List<PluginProvider> expProviders) throws Exception {
+        List<String> exp = toClasses(expProviders);
+
+        IgniteConfiguration cfg = getConfiguration();
+
+        cfg.setPluginProviders(cfgProviders);
+
+        IgniteEx ignite = startGrid(cfg);
+
+        List<String> providers = toClasses(ignite.context().plugins().allProviders());
+
+        assertEqualsCollections(exp, providers);
+    }
+
+    /**
+     * @param col Collection of plugin providers
+     */
+    @SuppressWarnings("rawtypes")
+    private static List<String> toClasses(Collection<PluginProvider> col) {
+        return col.stream().map(PluginProvider::name).collect(Collectors.toList());
+    }
+
+    /** Plugin with own message factory. */
+    private static class TestPlugin implements IgnitePlugin {
+    }
+
+    /** */
+    @SuppressWarnings("RedundantThrows")
+    public static class TestPluginProvider implements PluginProvider<TestPluginConfiguration> {
```

Comment:
Use AbstractTestPluginProvider to avoid boilerplate code, please.

Reply:
@ololo3000 Thanks. I'll do it in other change where some set of tests will be changed.